### PR TITLE
Generate DT_<COMPAT>_<INSTANCE>_<PROP> defines

### PR DESCRIPTION
**Pull request by galak**
Add support to generate defines of the form DT_<COMPAT>_<INSTANCE>_<PROP>.  The idea is that we can utilize this in drivers to remove the need for dts_fixup.h.  The <INSTANCE> value is determined by the script and starts at 0 and counts up for each instance of a given <COMPAT>.  

This PR is based on some other cleanup work (PR #12423).

### CODE REVIEW

**Comment by galak** [comment]
@anangl My intent would be to use this to replace the work you did with aliases for SoCs.

File: scripts/dts/extract/globals.py [hunk 1]
```diff
@@ -168,6 +172,9 @@ def find_node_by_path(nodes, path):
 
 
 def get_reduced(nodes, path):
+    if 'compat_count' not in get_reduced.__dict__:
```
Comment (anangl) [inline scripts/dts/extract/globals.py]: The name `compat_count` may be a bit misleading as the values stored here are the last assigned instance IDs, not numbers of occurrences of compats.

**Review by anangl — COMMENTED** [review]
I like this approach. We will no longer need to define aliases only to get rid of fixups. Thanks!

**Comment by erwango** [comment]
@galak, fyi, there is one case of fixup that won't be removed by this change:
https://github.com/zephyrproject-rtos/zephyr/blob/b7b02c7ec29ccf13b8e477352a9c61c758230dee/boards/arm/disco_l475_iot1/dts_fixup.h#L31

File: scripts/dts/extract/compatible.py [hunk 2]
```diff
@@ -54,6 +54,17 @@ def extract(self, node_address, prop, def_label):
                 }
                 insert_defs(node_address, load_defs, {})
 
+        # Generate defines of the form:
+        # #define DT_<COMPAT>_<INSTANCE ID> 1
```
[the remark this answers is not in the document]
Comment (erwango) [reply scripts/dts/extract/compatible.py]: I'm not actually sure, the bus label #define is used today as a compile switch to know how to compile driver for sensors compatible with multiple bus supported (say i2c or spi). As soon as one instance is there, "define should be generated so matching driver is compiled, whatever the instance in use (cf #12105).
This being said, this deserves a comment.

File: boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts [hunk 3]
```diff
@@ -58,9 +58,11 @@
 };
 
 &flexspi0 {
+	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	qspi0: qspi@60000000 {
-		/* ISSI IS25LP064A-JBLE */
-		reg = <0x60000000 0x800000>;
+		/* ISSI IS25WP064AJBLE */
+		compatible = "issi,is25wp064", "jedec,spi-nor";
```
Comment (pabigot) [inline boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts]: We don't have a `jedec,spi-nor` compatible yet; it's spelled `spi,flash` (supported by `drivers/flash/spi_nor`).  However I agree it should be `jedec,spi-nor` and in fact have support pending to be part of extending #11977 to include the `spi_nor` driver.

Do you want to leave it like this for now, or use `spi,flash` and I'll fix it later?

**Review by pabigot — DISMISSED** [review]
Generally looks good, need the instance added to the bus identifier and resolution of the compatible for spi flash.  May be more as I update #11977 and discover anything subtle.

File: scripts/dts/extract/compatible.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pabigot) [reply scripts/dts/extract/compatible.py]: It is possible to support multiple instances of the same sensor on different bus types, even if no in-tree code chooses to do so.  I wouldn't want the infrastructure to artificially reduce a system designer's options.

File: scripts/dts/extract/globals.py [hunk 4]
```diff
@@ -167,6 +167,9 @@ def find_node_by_path(nodes, path):
...
+        get_reduced.compat_count = {}
+
     # compress nodes list to nodes w/ paths, add interrupt parent
```
Comment (erwango) [inline scripts/dts/extract/globals.py]: This comment seems weird at that place now.

File: scripts/dts/extract/compatible.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (erwango) [reply scripts/dts/extract/compatible.py]: Sure, but this specifc bus tye #define was made to help compilation of the adequate variant of the driver. You need to compile sensor_spi.c whatever number of spi sensors are there. This is not instance dependant.

File: scripts/dts/extract/compatible.py [hunk 5]
```diff
@@ -54,6 +54,17 @@ def extract(self, node_address, prop, def_label):
...
+        instance = reduced[node_address]['instance_id']
+        for k in instance:
+            i = instance[k]
+            compat_instance = 'DT_' + convert_string_to_label(k) + '_' + str(i)
```
Comment (erwango) [inline scripts/dts/extract/compatible.py]: So do we want to always have _0 if only one instance is (and will ever be) available (eg for RTC)?
Or can we __also__ generate the alternate 'no index' #define when instance number is 0?
(This would spare the rework of sensors drivers)

**Review by erwango — DISMISSED** [review]

File: scripts/dts/extract/compatible.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (erwango) [reply scripts/dts/extract/compatible.py]: Sure, but this specifc bus tye #define was made to help compilation of the adequate variant of the driver. You need to compile sensor_spi.c whatever number of spi sensors are there. This is not instance dependant.

[the remark this answers is not in the document]
Comment (pabigot) [reply scripts/dts/extract/compatible.py]: That's a good point, but the driver still needs to know whether DT_COMPAT_INSTANCE_ should be defined to use the I2C or the SPI API.  So if the unqualified `DT_COMPAT_BUS_type` is used to enable/disable the bus-specific driver, there still needs to be an instance-qualified version to select the bus for that instance.

[the remark this answers is not in the document]
Comment (erwango) [reply scripts/dts/extract/compatible.py]: > That's a good point, but the driver still needs to know whether DT_COMPAT_INSTANCE_ should be defined to use the I2C or the SPI API. 

Right, but we don't have this code there yet, so I agree information will be needed, but we don't know how.

> So if the unqualified `DT_COMPAT_BUS_type` is used to enable/disable the bus-specific driver, there still needs to be an instance-qualified version to select the bus for that instance.

Then we might need to generate another more explicit define:
DT_<COMPAT>_<INSTANCE>_BUS_<BUS_TYPE>

[the remark this answers is not in the document]
Comment (pabigot) [reply scripts/dts/extract/compatible.py]: I don't understand the second part of your response; for the first if we generate both the instance-qualified and non-instance-qualified bus defines then it's pretty likely the information we need will be available and we can make something work.  If we don't have the instance-qualified one we definitely can't make it work.

[the remark this answers is not in the document]
Comment (erwango) [reply scripts/dts/extract/compatible.py]: Yeah, sorry, I was betrailled by markdown and I was maybe not clear enough.
So I meant rather than:
DT_COMPAT_BUS_\<TYPE\>_INSTANCE
generate:
DT_COMPAT_INSTANCE_BUS_\<TYPE\>

[the remark this answers is not in the document]
Comment (erwango) [reply scripts/dts/extract/compatible.py]: > Is that OK?

Yep!

File: boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts (hunk 3 above)
Comment (galak) [reply boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts]: Leave it for now, there isn't proper support for the 'issi' flash so just wanted the fix to the structure of the nodes.

File: scripts/dts/extract/compatible.py (hunk 5 above)
Comment (galak) [reply scripts/dts/extract/compatible.py]: I always want _0.  I'd rather the naming convention not try to be overly smart.  The extra '_0_' seems minor for the cases that only one instance will ever exist.

File: scripts/dts/extract/globals.py (hunk 1 above)
Comment (galak) [reply scripts/dts/extract/globals.py]: Suggestions on name?

File: scripts/dts/extract/compatible.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (galak) [reply scripts/dts/extract/compatible.py]: Done

File: scripts/dts/extract_dts_includes.py [hunk 6]
```diff
@@ -294,6 +296,7 @@ def extract_single(node_address, prop, key, def_label):
             label = def_label + '_' + k
             if isinstance(p, str):
                 p = "\"" + p + "\""
+            add_compat_alias(node_address, k, label, prop_alias)
```
[the remark this answers is not in the document]
Comment (galak) [reply scripts/dts/extract_dts_includes.py]: Why do you think 'k' should be decorated?

File: scripts/dts/extract/globals.py (hunk 4 above)
Comment (galak) [reply scripts/dts/extract/globals.py]: moved it back up.

File: scripts/dts/extract/globals.py (hunk 1 above)
Comment (galak) [reply scripts/dts/extract/globals.py]: Unfortunately as best I can tell there isn't a dts that hits this code path, so not sure what its trying to do anymore.

File: scripts/dts/extract_dts_includes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (galak) [reply scripts/dts/extract_dts_includes.py]: Unfortunately as best I can tell there isn't a dts that hits this code path, so not sure what its trying to do anymore.

[the remark this answers is not in the document]
Comment (anangl) [reply scripts/dts/extract_dts_includes.py]: > Why do you think 'k' should be decorated?

Because otherwise the same alias would be added multiple times with different values, and effectively, after filtering, only the one for the last position from the list would get generated.

> Unfortunately as best I can tell there isn't a dts that hits this code path, so not sure what its trying to do anymore.

Yes, it is not currently used. And I have no clue what case this path was intended to handle. I only added the comment because this code was updated in this PR.

File: scripts/dts/extract/globals.py (hunk 1 above)
Comment (anangl) [reply scripts/dts/extract/globals.py]: `last_used_id` maybe?

Comment (galak) [reply scripts/dts/extract/globals.py]: Done

**Comment by galak** [comment]
I tried to convert some SoC drivers to use this, and there are still issues that cause this not to fully work for all SoCs.  There are various cases that depend on know which specific instance of a device you are talking to.  For example pinmux control.

This change should work for I2C & SPI devices which is the biggest area of focus.

**Comment by MaureenHelm** [comment]
How will this work with the way we conditionally enable driver instances in Kconfig? For example, `drivers/gpio/gpio_mcux_igpio.c` has instances conditioned by `#ifdef CONFIG_GPIO_MCUX_IGPIO_1`.
https://github.com/zephyrproject-rtos/zephyr/blob/c2c9265b7ddbd9300e829438e9e78d6ee4b6177d/drivers/gpio/gpio_mcux_igpio.c#L173-L202

If I replace the base address macro, which currently comes from a fixup, with the newly generated one for instance 0, there's no guarantee it's the right instance. For mimxrt1020_evk, it ends up being the address for instance 5.

I could change the instance to be conditioned by `#ifdef DT_NXP_IMX_GPIO_0`, but now the only way to turn the right instances on/off is with a dts overlay. Do Kconfig instance enables go away as a consequence?

**Comment by galak** [comment]
> I could change the instance to be conditioned by `#ifdef DT_NXP_IMX_GPIO_0`, but now the only way to turn the right instances on/off is with a dts overlay. Do Kconfig instance enables go away as a consequence?

Yes Kconfig instances need to go away as well.  Take a look at what I have in PR #12442

**Comment by mnkp** [comment]
> There are various cases that depend on know which specific instance of a device you are talking to. For example pinmux control.

These are kind of issues we're going to run into when we implement automatic instance generation so sooner or later we need to solve them anyway. Though it's going to be a bit of work.

There are some SoC families, e.g. Atmel SAM and SiLabs which use different approach to configure GPIO pins. Rather than use pinmux the pin configuration happens when the device instance (uart, i2c, etc.) is initialized. This means that every instance has to know which GPIO pins it is using and the allowed choices are defined by the datasheet. Currently this information comes from a pinmap.h file. To be compatible with `DT_<COMPAT>_<INSTANCE>_<PROP>` approach I reckon it would be best to embed pinmap information directly in DTS. After all that's what DTS is meant for. At least for the SiLabs family this should render pin configuration much more flexible than it is now.

**Comment by galak** [comment]
> These are kind of issues we're going to run into when we implement automatic instance generation so sooner or later we need to solve them anyway. Though it's going to be a bit of work.

Agreed, its useful to be able to see these issues before we get there and can work on cleaning them up.
 
> There are some SoC families, e.g. Atmel SAM and SiLabs which use different approach to configure GPIO pins. Rather than use pinmux the pin configuration happens when the device instance (uart, i2c, etc.) is initialized. This means that every instance has to know which GPIO pins it is using and the allowed choices are defined by the datasheet. Currently this information comes from a pinmap.h file. To be compatible with `DT_<COMPAT>_<INSTANCE>_<PROP>` approach I reckon it would be best to embed pinmap information directly in DTS. After all that's what DTS is meant for. At least for the SiLabs family this should render pin configuration much more flexible than it is now.

Agreed, this should be the case across the board for pinmux.

File: scripts/dts/extract/interrupts.py [hunk 7]
```diff
@@ -89,6 +96,7 @@ def extract(self, node_address, prop, names, def_label):
                         l_fqn,
                         prop_alias)
 
+
```
Comment (pabigot) [inline scripts/dts/extract/interrupts.py]: too many empty lines?

**Review by pabigot — APPROVED** [review]
This looks good to me, with one formatting anomaly that I don't consider a blocker.

I've updated what was in #11977 to be rebased on this, visible at https://github.com/pabigot/zephyr/tree/rb/11977-on-12426.  The main lesson learned is that there are some non-standard uses where the lack of being able to reliably reference material associated with a specific device tree node may cause some pain.  This is because the generated instance number may change if the bindings file is reorganized, so the nodes identified by `DT_COMPAT_0_LABEL` and `DT_COMPAT_1_LABEL` may change, as noted by @MaureenHelm and others during previous review and discussions.

In the case of the i2c slave eeprom it's because that test does manipulation of the eeprom through its parent bus, which can't be identified at build time using the label that's attached to the eeprom device.  In the rebase I [noted in a comment](https://github.com/pabigot/zephyr/blob/120554df98142a4fd7a4a1df16d85c57f45a456f/tests/drivers/i2c/i2c_slave_api/src/main.c#L157) why that was done, and I switched the sht3xd example to using the literal device name string which will be the standard going forward.

In other words there's some evidence that being able to access node properties through `DT_ALIAS_foo` macros (not currently in tree) or a full device tree path (also not supported) may still be useful.  Once the `DT_COMPAT_INSTANCE_foo` material is in tree I would like to see `DT_COMPAT_ALIAS_foo` replaced by `DT_ALIAS_foo` to support these situations and to replace the old alias `ALIAS_foo` defines that are not namespaced but are generated by default given `--old-alias-names` in `dts.cmake`.

That need has no impact on this PR, which I think is a nice step forward.

File: scripts/dts/extract/interrupts.py (hunk 7 above)
Comment (galak) [reply scripts/dts/extract/interrupts.py]: Removed.

**Review by anangl — APPROVED** [review]

File: scripts/dts/extract/globals.py [hunk 8]
```diff
@@ -276,6 +297,14 @@ def enable_old_alias_names(enable):
     global old_alias_names
     old_alias_names = enable
 
+def add_compat_alias(node_address, label_postfix, label, prop_aliases):
```
Comment (mnkp) [inline scripts/dts/extract/globals.py]: Using 'alias' in the function name is a bit confusing as it suggests to the reader that the function has something to do with DTS aliases node.

**Review by mnkp — APPROVED** [review]
I didn't analyze the python code, however I tested this PR by generating defines for Atmel and Silabs SoCs. It works very well.

**Comment by carlescufi** [comment]
@galak @MaureenHelm is this ready for merging?

**Comment by galak** [comment]
> @galak @MaureenHelm is this ready for merging?

Yes, just merged it.
